hashprobe: remove dependency on runtime.aeskeysched

asubiotto opened the pull request:
This is used to seed the AES hash function. This commit reimplements aeskeysched locally to the package to avoid a dependency on runtime internals. In Go1.23 (yet to be released), runtime internals will require an explicit go:linkname directive to allow external packages to use them in order to lock down dependencies on runtime internals. runtime.aeskeysched is not currently exported so parquet-go builds are broken on Go1.23.

We could create a CL to add go:linkname to aeskeysched, but I don't think there's a good reason we can't remove the runtime dependency here.

asubiotto commented:
I'm going to rely on CI to verify correctness here so this might not be ready for a review yet.

asubiotto commented:
Looks like this works althought the tests need to be updated since our aeshash no longer uses the same seed as the `memhash` runtime functions. What would some acceptable tests be?

achille-roussel commented on hashprobe/aeshash/aeshash_amd64.go:
```diff
@@ -3,10 +3,25 @@
+func init() {
+	key := (*[hashRandomBytes / 8]uint64)(unsafe.Pointer(&aeskeysched))
+	for i := range key {
+		key[i] = rand.Uint64()
```
An issue with `rand` here is that it will be called before the program usually has an opportunity to set the random seed, which means the `aeskeysched` value will always be the same.

It looks like the Go runtime uses [chacha8](https://github.com/golang/go/blob/master/src/internal/chacha8rand/chacha8.go#L20) for these random values; maybe we can seed from `crypto/rand` by default, and assign the `aeskeysched` varaible to a known value in tests so we can have deterministic outputs?

asubiotto replied:
I'm not sure it's necessary to do anything else here because we're using the global random number generator (which is seeded by the runtime, not the program).

achille-roussel replied:
> If Seed is not called, the generator is seeded randomly at program startup.

Prior to Go 1.20, the generator was seeded like Seed(1) at program startup. To force the old behavior, call Seed(1) at program startup. Alternately, set GODEBUG=randautoseed=0 in the environment before making any calls to functions in this package.

I didn't know this was changed in Go 1.20 👍

achille-roussel reviewed (commented):
For tests, the ideal

asubiotto commented on hashprobe/aeshash/aeshash_amd64.go:
```diff
@@ -3,10 +3,25 @@
+	"github.com/parquet-go/parquet-go/sparse"
 )
 
+const hashRandomBytes = unsafe.Sizeof(uintptr(0)) / 4 * 64
```
Done.

asubiotto commented:
I'll double check aeskeysched is different across runs as written and have modified the tests to use a deterministically set aeskeysched and will update to compare to constants.

asubiotto commented:
Not very scientific, but I just panicked with the `aeskeyched` value and ensured it wasn't the same across CI runs:
![image](https://github.com/parquet-go/parquet-go/assets/10560359/44eb32dd-d22e-4b8f-833e-6495bb0edcc2)
![image](https://github.com/parquet-go/parquet-go/assets/10560359/4ad15099-aa28-4b6f-a0e4-9b9dc8c90599)

asubiotto commented:
OK, this should now be good to go. I added a `testingInitAesKeySched` function to deterministically initialize the seed in tests. It is defined as a noop in purego to avoid compilation errors. I just added it to the existing files even though it's a testing-only function to avoid complexity but I can also create `_test.go` files with compilation directives if we want to do that.

achille-roussel commented on hashprobe/aeshash/aeshash_amd64.go:
```diff
@@ -3,10 +3,34 @@
+	}
+}
+
+// testingInitAesKeySched initializes the aes seed to a deterministic value for
+// test purposes.
+func testingInitAesKeySched() {
+	for i := range aeskeysched {
+		aeskeysched[i] = byte(i)
+	}
+}
```
One way Go typically avoids having testing code in the package would be to make a `export_amd64_test.go` file that looks like this:

```go
//go:build amd64
package aeshash

func TestingInitAeskeysched() {
	for i := range aeskeysched {
		aeskeysched[i] = byte(i)
	}
}
```

then you can call `aeshash.TestingInitAeskeysched` from tests because the `*_test` files are included when building tests, and the function is part of the `aeshash` package.

asubiotto replied:
Yes, sorry, this is what I meant with "create _test.go files with compilation directives". Will do this then.

achille-roussel commented on hashprobe/aeshash/aeshash_test.go:
```diff
@@ -4,43 +4,19 @@ import (
 
-	h0 := memhash32(42, 1)
-	h1 := Hash32(42, 1)
+	testingInitAesKeySched()
```
Could we call this from an `init` function in the test file so we don't have to repeat it in each test?

asubiotto replied:
I don't think we have any guarantees that the testing `init` function will be called before the other `init` function which sets `aeskeysched` too which is why I did it this way even though I don't like it either. Will think if there's another way to do it.

achille-roussel replied:
Ah! Because the tests are in the same package. 

We could enforce it by moving tests to a aeshash_test package but that seems a bit overkill, the current approach is good enough 👍

asubiotto replied:
I just modified it to check `aeskeysched` before initialization in the `amd64.go` file which I think should work regardless of `init` order.

asubiotto commented:
Changed to ensure that `aeskeysched` is unset in the amd64 init function before initializing to be able to deterministically initialize `aeskeysched`. Since `init`  functions are called sequentially from the same goroutine this should be valid regardless of whether that init function is called before or after the testing init function which sets or overwrites `aeskeysched`
